code/modules/vehicles/mecha/combat/justice.dm

```diff
@@ -0,0 +1,455 @@
+	SIGNAL_HANDLER
+
+	if(!isliving(target))
+		return
+	var/mob/living/live_or_dead = target
+	if(live_or_dead.stat < UNCONSCIOUS && live_or_dead.getStaminaLoss() < 100)
+		return FALSE
+	if(charge_action?.on)
+		return FALSE
+	if(stealth_action?.on)
+		return FALSE
+	say(pick("Take my Justice-Slash!", "A falling leaf...", "Justice is quite a lonely path"), forced = "Justice Mech")
+	playsound(src, 'sound/mecha/mech_stealth_pre_attack.ogg', 75, FALSE)
+	addtimer(CALLBACK(src, PROC_REF(finish_him), pilot, live_or_dead), 1 SECONDS)
+	return TRUE
+
+/**
+ * ## finish_him
+ *
+ * Target's head is cut off (if it has one),
+ * Otherwise it deals 100 damage.
+ * Attack from invisability and charged attack have higher priority.
+ * Arguments:
+ * * finisher - Mech pilot who makes an attack.
+ * * him - Target at which the mech makes an attack.
+ */
+/obj/vehicle/sealed/mecha/justice/proc/finish_him(mob/finisher, mob/living/him)
+	/// turf where we end attack
+	var/turf/finish_turf = get_step(him, get_dir(finisher, him))
+	/// turf where we start attack
+	var/turf/for_line_turf = get_turf(finisher)
+	var/obj/item/bodypart/in_your_head = him.get_bodypart(BODY_ZONE_HEAD)
+	if(in_your_head)
+		in_your_head.dismember(BRUTE)
+	else
+		him.apply_damage(100, BRUTE)
+	playsound(src, brute_attack_sound, 75, FALSE)
+	for_line_turf.Beam(src, icon_state = "mech_charge", time = 8)
+	forceMove(finish_turf)
+
+/obj/vehicle/sealed/mecha/justice/melee_attack_effect(mob/living/victim, heavy)
+	if(!heavy)
+		victim.Knockdown(40)
+		return
+	var/obj/item/bodypart/cut_bodypart = victim.get_bodypart(pick(BODY_ZONE_R_ARM, BODY_ZONE_R_LEG, BODY_ZONE_L_ARM, BODY_ZONE_L_LEG))
+	if(prob(DISMEMBER_CUALITY_HIGTH))
+		cut_bodypart.dismember(BRUTE)
+
+
+/obj/vehicle/sealed/mecha/justice/mob_exit(mob/M, silent, randomstep, forced)
+	. = ..()
+	if(!stealth_action?.on)
+		return
+	animate(src, alpha = 255, time = 0.5 SECONDS)
+	playsound(src, 'sound/mecha/mech_stealth_effect.ogg' , 75, FALSE)
+
+/obj/vehicle/sealed/mecha/justice/Bump(atom/obstacle)
+	. = ..()
+	if(!isliving(obstacle))
+		return
+	if(!stealth_action?.on)
+		return
+	animate(src, alpha = 255, time = 0.5 SECONDS)
+	playsound(src, 'sound/mecha/mech_stealth_effect.ogg' , 75, FALSE)
+
+/obj/vehicle/sealed/mecha/justice/Bumped(atom/movable/bumped_atom)
+	. = ..()
+	if(!isliving(bumped_atom))
+		return
+	if(!stealth_action?.on)
+		return
+	animate(src, alpha = 255, time = 0.5 SECONDS)
+	playsound(src, 'sound/mecha/mech_stealth_effect.ogg' , 75, FALSE)
+
+/obj/vehicle/sealed/mecha/justice/take_damage(damage_amount, damage_type, damage_flag, sound_effect, attack_dir, armour_penetration)
+	if(stealth_action?.on)
+		animate(src, alpha = 255, time = 0.5 SECONDS)
+		playsound(src, 'sound/mecha/mech_stealth_effect.ogg' , 75, FALSE)
+	if(LAZYLEN(occupants))
+		if(prob(60))
+			new /obj/effect/temp_visual/mech_sparks(get_turf(src))
+			playsound(src, 'sound/mecha/mech_stealth_effect.ogg' , 75, FALSE)
+			return
```

Comment:
And this.
Everything related to the ability should be in the ability's code, not the mech's.

Reply:
I left variables where I can’t do it through signals